<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Bioconductor 2020 Annotation Workshop • Bioc2020Anno</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Bioconductor 2020 Annotation Workshop">
<meta property="og:description" content="Bioc2020Anno">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Bioc2020Anno</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.6.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/AnnotationWorkshop.html">Bioconductor 2020 Annotation Workshop</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jmacdon/Bioc2020Anno">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="AnnotationWorkshop_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Bioconductor 2020 Annotation Workshop</h1>
            
      
      
      <div class="hidden name"><code>AnnotationWorkshop.Rmd</code></div>

    </div>

    
    
<div id="introduction-to-bioconductor-annotation-resources" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction-to-bioconductor-annotation-resources" class="anchor"></a>Introduction to Bioconductor Annotation Resources</h1>
<p>Bioc 2020: 29-31 July</p>
<div id="instructors" class="section level2">
<h2 class="hasAnchor">
<a href="#instructors" class="anchor"></a>Instructors</h2>
<ul>
<li>James W. MacDonald (<a href="mailto:jmacdon@umich.edu" class="email">jmacdon@umich.edu</a>)</li>
<li>Lori Shepherd (<a href="mailto:lori.shepherd@roswellpark.org" class="email">lori.shepherd@roswellpark.org</a>)</li>
</ul>
</div>
<div id="workshop-description" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-description" class="anchor"></a>Workshop Description</h2>
<p>There are various annotation packages provided by the Bioconductor project that can be used to incorporate additional information to results from high-throughput experiments. This can be as simple as mapping Ensembl IDs to corresponding HUGO gene symbols, to much more complex queries involving multiple data sources. In this workshop we will cover the various classes of annotation packages, what they contain, and how to use them efficiently.</p>
<div id="prerequisites" class="section level3">
<h3 class="hasAnchor">
<a href="#prerequisites" class="anchor"></a>Prerequisites</h3>
<ul>
<li>Basic knowledge of R syntax</li>
<li>Basic understanding of the various annotation sources (NCBI, EBI/EMBL)</li>
</ul>
<p>Useful background reading</p>
<ul>
<li>The <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/AnnotationDbi/inst/doc/IntroToAnnotationPackages.pdf">AnnotationDbi</a> vignette.</li>
<li>The <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/biomaRt.html">biomaRt</a> vignette.</li>
<li>The <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/GenomicFeatures/inst/doc/GenomicFeatures.pdf">GenomicFeatures</a> vignette.</li>
</ul>
</div>
<div id="workshop-participation" class="section level3">
<h3 class="hasAnchor">
<a href="#workshop-participation" class="anchor"></a>Workshop Participation</h3>
<p>After each type of annotation package is introduced, students will be given the opportunity to practice making their own queries.</p>
</div>
<div id="r-bioconductor-packages-used" class="section level3">
<h3 class="hasAnchor">
<a href="#r-bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h3>
<ul>
<li>AnnotationDbi</li>
<li>AnnotationHub</li>
<li>BSgenome</li>
<li>biomaRt</li>
<li>ensembldb</li>
<li>org.Hs.eg.db</li>
<li>TxDb.Hsapiens.UCSC.hg19.knownGene</li>
<li>EnsDb.Hsapiens.v79</li>
<li>EnsDb.Mmusculus.v79</li>
<li>Homo.sapiens</li>
<li>BSgenome.Hsapiens.UCSC.hg19</li>
<li>hugene20sttranscriptcluster.db</li>
</ul>
</div>
</div>
<div id="workshop-goals-and-objectives" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-goals-and-objectives" class="anchor"></a>Workshop goals and objectives</h2>
<p>Annotating data is a complex task. For any high-throughput experiment the analyst usually starts with a set of identifiers for each thing that was measured, and in order to make the results useful to collaborators these identifiers need to be mapped to other identifiers that are either more familiar to collaborators, or that can be used for further analyses. As an example, RNA-Seq data may only have Entrez Gene IDs for each gene measured, and as part of the output you may want to include the gene symbols, which are more likely to be familiar to a Biologist.</p>
<div id="learning-goals" class="section level3">
<h3 class="hasAnchor">
<a href="#learning-goals" class="anchor"></a>Learning goals</h3>
<ul>
<li>Understand what sort of annotation data are available</li>
<li>Understand the difference between annotation sources (NCBI and EBI/EMBL)</li>
<li>Gain familiarity with the various ways to query annotation packages</li>
<li>Get some practice making queries</li>
</ul>
</div>
<div id="learning-objectives" class="section level3">
<h3 class="hasAnchor">
<a href="#learning-objectives" class="anchor"></a>Learning objectives</h3>
<ul>
<li>Be able to use select and mapIds to map between identifiers</li>
<li>Be able to extract data from TxDb and EnsDb packages</li>
<li>Be able to make queries using biomaRt</li>
<li>Extract and utilize various data from AnnotationHub</li>
</ul>
</div>
</div>
<div id="annotation-workshop" class="section level2">
<h2 class="hasAnchor">
<a href="#annotation-workshop" class="anchor"></a>Annotation Workshop</h2>
</div>
<div id="what-do-we-mean-by-annotation" class="section level2">
<h2 class="hasAnchor">
<a href="#what-do-we-mean-by-annotation" class="anchor"></a>What do we mean by annotation?</h2>
<p>Map a known ID to other functional or positional information</p>
<p><img src="mapping.png" width="75%"></p>
</div>
<div id="specific-goal" class="section level2">
<h2 class="hasAnchor">
<a href="#specific-goal" class="anchor"></a>Specific goal</h2>
<p>We have data and statistics, and we want to add other useful information</p>
<p><img src="boxes.png" width="75%"></p>
<p>The end result might be as simple as a data.frame or HTML table, or as complex as a <code>RangedSummarizedExperiment</code></p>
</div>
<div id="data-containers" class="section level2">
<h2 class="hasAnchor">
<a href="#data-containers" class="anchor"></a>Data containers</h2>
<p><img src="eset.png" width="75%"></p>
</div>
<div id="expressionset" class="section level2">
<h2 class="hasAnchor">
<a href="#expressionset" class="anchor"></a>ExpressionSet</h2>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="kw">package</span> <span class="kw">=</span> <span class="st">"Bioc2020Anno"</span>, <span class="st">"extdata"</span>), <span class="st">"eset.Rdata"</span>))
<span class="no">eset</span></pre></body></html></div>
<pre><code>## ExpressionSet (storageMode: lockedEnvironment)
## assayData: 33552 features, 6 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: GSM2194079 GSM2194080 ... GSM2194084 (6 total)
##   varLabels: title characteristics_ch1.1
##   varMetadata: labelDescription
## featureData
##   featureNames: 16657436 16657440 ... 17118478 (33552 total)
##   fvarLabels: PROBEID ENTREZID SYMBOL GENENAME
##   fvarMetadata: labelDescription
## experimentData: use 'experimentData(object)'
## Annotation: pd.hugene.2.0.st</code></pre>
</div>
<div id="expressionset-continued" class="section level2">
<h2 class="hasAnchor">
<a href="#expressionset-continued" class="anchor"></a>ExpressionSet (continued)</h2>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu">exprs</span>(<span class="no">eset</span>))</pre></body></html></div>
<pre><code>##          GSM2194079 GSM2194080 GSM2194081 GSM2194082 GSM2194083 GSM2194084
## 16657436   8.505158   9.046577   8.382674   9.115481   8.715343   8.566301
## 16657440   7.948860   8.191222   7.901911   8.459781   8.191793   8.219658
## 16657450  10.932934  11.228553  10.948120  11.462231  11.300046  11.300886
## 16657469   9.172462   9.344630   9.193450   9.465584   9.464020   9.135715
## 16657473   6.222049   6.551035   6.000246   6.398798   5.892654   5.592125
## 16657476   8.514300   8.474073   8.407196   8.811238   8.780833   8.874606</code></pre>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu">pData</span>(<span class="fu">phenoData</span>(<span class="no">eset</span>)))</pre></body></html></div>
<pre><code>##                          title characteristics_ch1.1
## GSM2194079   SW620-miR625-rep1     shRNA: miR-625-3p
## GSM2194080   SW620-miR625-rep2     shRNA: miR-625-3p
## GSM2194081   SW620-miR625-rep3     shRNA: miR-625-3p
## GSM2194082 SW620-scramble-rep1       shRNA: scramble
## GSM2194083 SW620-scramble-rep2       shRNA: scramble
## GSM2194084 SW620-scramble-rep3       shRNA: scramble</code></pre>
</div>
<div id="expressionset-continued-1" class="section level2">
<h2 class="hasAnchor">
<a href="#expressionset-continued-1" class="anchor"></a>ExpressionSet (continued)</h2>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu">pData</span>(<span class="fu">featureData</span>(<span class="no">eset</span>)))</pre></body></html></div>
<pre><code>##           PROBEID  ENTREZID      SYMBOL
## 16657436 16657436     84771     DDX11L2
## 16657440 16657440 100302278   MIR1302-2
## 16657450 16657450    402483   LINC01000
## 16657469 16657469    140849 LINC00266-1
## 16657473 16657473    729759      OR4F29
## 16657476 16657476    388574   RPL23AP87
##                                                   GENENAME
## 16657436                     DEAD/H-box helicase 11 like 2
## 16657440                                   microRNA 1302-2
## 16657450       long intergenic non-protein coding RNA 1000
## 16657469      long intergenic non-protein coding RNA 266-1
## 16657473 olfactory receptor family 4 subfamily F member 29
## 16657476              ribosomal protein L23a pseudogene 87</code></pre>
</div>
<div id="bioc-containers-vs-basic-structures" class="section level2">
<h2 class="hasAnchor">
<a href="#bioc-containers-vs-basic-structures" class="anchor"></a>BioC containers vs basic structures</h2>
<div id="pros" class="section level3">
<h3 class="hasAnchor">
<a href="#pros" class="anchor"></a>Pros</h3>
<ul>
<li><p>Validity checking</p></li>
<li><p>Subsetting</p></li>
<li><p>Function dispatch</p></li>
<li><p>Automatic behaviors</p></li>
</ul>
</div>
<div id="cons" class="section level3">
<h3 class="hasAnchor">
<a href="#cons" class="anchor"></a>Cons</h3>
<ul>
<li><p>Difficult to create</p></li>
<li><p>Cumbersome to extract data by hand</p></li>
<li><p>Useful only within R</p></li>
</ul>
</div>
</div>
<div id="annotation-sources" class="section level2">
<h2 class="hasAnchor">
<a href="#annotation-sources" class="anchor"></a>Annotation sources</h2>
<table class="table">
<thead><tr class="header">
<th align="left">Package type</th>
<th align="left">Example</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ChipDb</td>
<td align="left">hugene20sttranscriptcluster.db</td>
</tr>
<tr class="even">
<td align="left">OrgDb</td>
<td align="left">org.Hs.eg.db</td>
</tr>
<tr class="odd">
<td align="left">TxDb/EnsDb</td>
<td align="left">TxDb.Hsapiens.UCSC.hg19.knownGene; EnsDb.Hsapiens.v75</td>
</tr>
<tr class="even">
<td align="left">OrganismDb</td>
<td align="left">Homo.sapiens</td>
</tr>
<tr class="odd">
<td align="left">BSgenome</td>
<td align="left">BSgenome.Hsapiens.UCSC.hg19</td>
</tr>
<tr class="even">
<td align="left">Others</td>
<td align="left">GO.db; KEGG.db</td>
</tr>
<tr class="odd">
<td align="left">AnnotationHub</td>
<td align="left">Online resource</td>
</tr>
<tr class="even">
<td align="left">biomaRt</td>
<td align="left">Online resource</td>
</tr>
</tbody>
</table>
</div>
<div id="interacting-with-annodb-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#interacting-with-annodb-packages" class="anchor"></a>Interacting with AnnoDb packages</h2>
<p>The main function is <code>select</code>:</p>
<p>select(<em>annopkg</em>, <em>keys</em>, <em>columns</em>, <em>keytype</em>)</p>
<p>Where</p>
<ul>
<li><p>annopkg is the annotation package</p></li>
<li><p>keys are the IDs that we <strong>know</strong></p></li>
<li><p>columns are the values we <strong>want</strong></p></li>
<li>
<p>keytype is the type of key used</p>
<ul>
<li>if the keytype is the <strong>central</strong> key, it can remain unspecified</li>
</ul>
</li>
</ul>
</div>
<div id="simple-example" class="section level2">
<h2 class="hasAnchor">
<a href="#simple-example" class="anchor"></a>Simple example</h2>
<p>Say we have analyzed data from an Affymetrix Human Gene ST 2.0 array and want to know what the genes are. For purposes of this lab, we just select some IDs at random.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">hugene20sttranscriptcluster.db</span>)
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">12345</span>)
<span class="no">ids</span> <span class="kw">&lt;-</span> <span class="fu">featureNames</span>(<span class="no">eset</span>)[<span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fl">1</span>:<span class="fl">25000</span>, <span class="fl">5</span>)]
<span class="no">ids</span></pre></body></html></div>
<pre><code>## [1] "16858710" "17001289" "16894087" "16813565" "16951883"</code></pre>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="fu">select</span>(<span class="no">hugene20sttranscriptcluster.db</span>, <span class="no">ids</span>, <span class="st">"SYMBOL"</span>)</pre></body></html></div>
<pre><code>## 'select()' returned 1:1 mapping between keys and columns</code></pre>
<pre><code>##    PROBEID      SYMBOL
## 1 16858710        JUNB
## 2 17001289 PPP2R2B-IT1
## 3 16894087   LINC01247
## 4 16813565   NR2F2-AS1
## 5 16951883     OSBPL10</code></pre>
</div>
<div id="questions" class="section level2">
<h2 class="hasAnchor">
<a href="#questions" class="anchor"></a>Questions!</h2>
<p>How do you know what the central keys are?</p>
<ul>
<li><p>If it’s a ChipDb, the central key are the manufacturer’s probe IDs</p></li>
<li><p>It’s sometimes in the name - org.Hs.eg.db, where ‘eg’ means Entrez Gene ID</p></li>
<li><p>You can see examples using e.g., head(keys(<em>annopkg</em>)), and infer from that</p></li>
<li><p>But note that it’s never necessary to know the central key, as long as you specify the keytype</p></li>
</ul>
</div>
<div id="more-questions" class="section level2">
<h2 class="hasAnchor">
<a href="#more-questions" class="anchor"></a>More questions!</h2>
<p>What keytypes or columns are available for a given annotation package?</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="fu">keytypes</span>(<span class="no">hugene20sttranscriptcluster.db</span>)</pre></body></html></div>
<pre><code>##  [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"
##  [6] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL"  "GENENAME"    
## [11] "GO"           "GOALL"        "IPI"          "MAP"          "OMIM"        
## [16] "ONTOLOGY"     "ONTOLOGYALL"  "PATH"         "PFAM"         "PMID"        
## [21] "PROBEID"      "PROSITE"      "REFSEQ"       "SYMBOL"       "UCSCKG"      
## [26] "UNIGENE"      "UNIPROT"</code></pre>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="fu">columns</span>(<span class="no">hugene20sttranscriptcluster.db</span>)</pre></body></html></div>
<pre><code>##  [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"
##  [6] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL"  "GENENAME"    
## [11] "GO"           "GOALL"        "IPI"          "MAP"          "OMIM"        
## [16] "ONTOLOGY"     "ONTOLOGYALL"  "PATH"         "PFAM"         "PMID"        
## [21] "PROBEID"      "PROSITE"      "REFSEQ"       "SYMBOL"       "UCSCKG"      
## [26] "UNIGENE"      "UNIPROT"</code></pre>
</div>
<div id="another-example" class="section level2">
<h2 class="hasAnchor">
<a href="#another-example" class="anchor"></a>Another example</h2>
<p>There is one issue with <code>select</code> however.</p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="no">ids</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'16737401'</span>,<span class="st">'16657436'</span> ,<span class="st">'16678303'</span>)
<span class="fu">select</span>(<span class="no">hugene20sttranscriptcluster.db</span>, <span class="no">ids</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"SYMBOL"</span>,<span class="st">"MAP"</span>))</pre></body></html></div>
<pre><code>## 'select()' returned 1:many mapping between keys and columns</code></pre>
<pre><code>##     PROBEID       SYMBOL     MAP
## 1  16737401        TRAF6   11p12
## 2  16657436      DDX11L1 1p36.33
## 3  16657436 LOC102725121 1p36.33
## 4  16657436      DDX11L2  2q14.1
## 5  16657436      DDX11L9 15q26.3
## 6  16657436     DDX11L10 16p13.3
## 7  16657436      DDX11L5  9p24.3
## 8  16657436     DDX11L16    Xq28
## 9  16657436     DDX11L16    Yq12
## 10 16678303         ARF1 1q42.13</code></pre>
</div>
<div id="the-mapids-function" class="section level2">
<h2 class="hasAnchor">
<a href="#the-mapids-function" class="anchor"></a>The <code>mapIds</code> function</h2>
<p>An alternative to <code>select</code> is <code>mapIds</code>, which gives control of duplicates</p>
<ul>
<li>
<p>Same arguments as <code>select</code> with slight differences</p>
<ul>
<li><p>The columns argument can only specify one column</p></li>
<li><p>The keytype argument <strong>must</strong> be specified</p></li>
<li><p>An additional argument, multiVals used to control duplicates</p></li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="fu">mapIds</span>(<span class="no">hugene20sttranscriptcluster.db</span>, <span class="no">ids</span>, <span class="st">"SYMBOL"</span>, <span class="st">"PROBEID"</span>)</pre></body></html></div>
<pre><code>## 'select()' returned 1:many mapping between keys and columns</code></pre>
<pre><code>##  16737401  16657436  16678303 
##   "TRAF6" "DDX11L1"    "ARF1"</code></pre>
</div>
<div id="choices-for-multivals" class="section level2">
<h2 class="hasAnchor">
<a href="#choices-for-multivals" class="anchor"></a>Choices for multiVals</h2>
<p>Default is <code>first</code>, where we just choose the first of the duplicates. Other choices are <code>list</code>, <code>CharacterList</code>, <code>filter</code>, <code>asNA</code> or a user-specified function.</p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="fu">mapIds</span>(<span class="no">hugene20sttranscriptcluster.db</span>, <span class="no">ids</span>, <span class="st">"SYMBOL"</span>, <span class="st">"PROBEID"</span>, <span class="kw">multiVals</span> <span class="kw">=</span> <span class="st">"list"</span>)</pre></body></html></div>
<pre><code>## 'select()' returned 1:many mapping between keys and columns</code></pre>
<pre><code>## $`16737401`
## [1] "TRAF6"
## 
## $`16657436`
## [1] "DDX11L1"      "LOC102725121" "DDX11L2"      "DDX11L9"      "DDX11L10"    
## [6] "DDX11L5"      "DDX11L16"    
## 
## $`16678303`
## [1] "ARF1"</code></pre>
</div>
<div id="choices-for-multivals-continued" class="section level2">
<h2 class="hasAnchor">
<a href="#choices-for-multivals-continued" class="anchor"></a>Choices for multiVals (continued)</h2>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="fu">mapIds</span>(<span class="no">hugene20sttranscriptcluster.db</span>, <span class="no">ids</span>, <span class="st">"SYMBOL"</span>, <span class="st">"PROBEID"</span>, <span class="kw">multiVals</span> <span class="kw">=</span> <span class="st">"CharacterList"</span>)</pre></body></html></div>
<pre><code>## 'select()' returned 1:many mapping between keys and columns</code></pre>
<pre><code>## CharacterList of length 3
## [["16737401"]] TRAF6
## [["16657436"]] DDX11L1 LOC102725121 DDX11L2 DDX11L9 DDX11L10 DDX11L5 DDX11L16
## [["16678303"]] ARF1</code></pre>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="fu">mapIds</span>(<span class="no">hugene20sttranscriptcluster.db</span>, <span class="no">ids</span>, <span class="st">"SYMBOL"</span>, <span class="st">"PROBEID"</span>, <span class="kw">multiVals</span> <span class="kw">=</span> <span class="st">"filter"</span>)</pre></body></html></div>
<pre><code>## 'select()' returned 1:many mapping between keys and columns</code></pre>
<pre><code>## 16737401 16678303 
##  "TRAF6"   "ARF1"</code></pre>
<div class="sourceCode" id="cb33"><html><body><pre class="r"><span class="fu">mapIds</span>(<span class="no">hugene20sttranscriptcluster.db</span>, <span class="no">ids</span>, <span class="st">"SYMBOL"</span>, <span class="st">"PROBEID"</span>, <span class="kw">multiVals</span> <span class="kw">=</span> <span class="st">"asNA"</span>)</pre></body></html></div>
<pre><code>## 'select()' returned 1:many mapping between keys and columns</code></pre>
<pre><code>## 16737401 16657436 16678303 
##  "TRAF6"       NA   "ARF1"</code></pre>
</div>
<div id="chipdborgdb-questions" class="section level2">
<h2 class="hasAnchor">
<a href="#chipdborgdb-questions" class="anchor"></a>ChipDb/OrgDb questions</h2>
<p>Using either the hugene20sttranscriptcluster.db or org.Hs.eg.db package,</p>
<ul>
<li><p>What gene symbol corresponds to Entrez Gene ID 1000?</p></li>
<li><p>What is the Ensembl Gene ID for PPARG?</p></li>
<li><p>What is the UniProt ID for GAPDH?</p></li>
<li><p>How many of the probesets from the ExpressionSet (eset) we loaded map to a single gene? How many don’t map to a gene at all?</p></li>
</ul>
</div>
<div id="txdb-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#txdb-packages" class="anchor"></a>TxDb packages</h2>
<p>TxDb packages contain positional information; the contents can be inferred by the package name</p>
<p><strong>TxDb.Species.Source.Build.Table</strong></p>
<ul>
<li>
<p>TxDb.Hsapiens.UCSC.hg19.knownGene</p>
<ul>
<li><p><em>Homo sapiens</em></p></li>
<li><p>UCSC genome browser</p></li>
<li><p>hg19 (their version of GRCh37)</p></li>
<li><p>knownGene table</p></li>
</ul>
</li>
</ul>
<p>TxDb.Dmelanogaster.UCSC.dm3.ensGene TxDb.Athaliana.BioMart.plantsmart22</p>
</div>
<div id="ensdb-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#ensdb-packages" class="anchor"></a>EnsDb packages</h2>
<p>EnsDb packages are similar to TxDb packages, but based on Ensembl mappings</p>
<ul>
<li>EnsDb.Rnorvegicus.v79</li>
<li>EnsDb.Hsapiens.v79</li>
<li>EnsDb.Mmusculus.v79</li>
<li>EnsDb.Rnorvegicus.v79</li>
</ul>
</div>
<div id="transcript-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#transcript-packages" class="anchor"></a>Transcript packages</h2>
<p>As with ChipDb and OrgDb packages, <code>select</code> and <code>mapIds</code> can be used to make queries</p>
<div class="sourceCode" id="cb36"><html><body><pre class="r"><span class="fu">select</span>(<span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"1"</span>,<span class="st">"10"</span>),
       <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"TXNAME"</span>,<span class="st">"TXCHROM"</span>,<span class="st">"TXSTART"</span>,<span class="st">"TXEND"</span>), <span class="st">"GENEID"</span>)</pre></body></html></div>
<pre><code>## 'select()' returned 1:many mapping between keys and columns</code></pre>
<pre><code>##   GENEID     TXNAME TXCHROM  TXSTART    TXEND
## 1      1 uc002qsd.4   chr19 58858172 58864865
## 2      1 uc002qsf.2   chr19 58859832 58874214
## 3     10 uc003wyw.1    chr8 18248755 18258723</code></pre>
<div class="sourceCode" id="cb39"><html><body><pre class="r"><span class="fu">select</span>(<span class="no">EnsDb.Hsapiens.v79</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"1"</span>, <span class="st">"10"</span>),
       <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"GENEID"</span>,<span class="st">"GENENAME"</span>,<span class="st">"SEQNAME"</span>,<span class="st">"GENESEQSTART"</span>,<span class="st">"GENESEQEND"</span>), <span class="st">"ENTREZID"</span>)</pre></body></html></div>
<pre><code>##   ENTREZID          GENEID GENENAME SEQNAME GENESEQSTART GENESEQEND
## 1        1 ENSG00000121410     A1BG      19     58345178   58353499
## 2       10 ENSG00000156006     NAT2       8     18391245   18401218</code></pre>
<p>But this is not how one normally uses them…</p>
</div>
<div id="granges" class="section level2">
<h2 class="hasAnchor">
<a href="#granges" class="anchor"></a>GRanges</h2>
<p>The normal use case for transcript packages is to extract positional information into a <code>GRanges</code> or <code>GRangesList</code> object. An example is the genomic position of all genes:</p>
<div class="sourceCode" id="cb41"><html><body><pre class="r"><span class="no">gns</span> <span class="kw">&lt;-</span> <span class="fu">genes</span>(<span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>)</pre></body></html></div>
<pre><code>##   403 genes were dropped because they have exons located on both strands
##   of the same reference sequence or on more than one reference sequence,
##   so cannot be represented by a single genomic range.
##   Use 'single.strand.genes.only=FALSE' to get all the genes in a
##   GRangesList object, or use suppressMessages() to suppress this message.</code></pre>
<div class="sourceCode" id="cb43"><html><body><pre class="r"><span class="no">gns</span></pre></body></html></div>
<pre><code>## GRanges object with 23056 ranges and 1 metadata column:
##         seqnames              ranges strand |     gene_id
##            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##       1    chr19   58858172-58874214      - |           1
##      10     chr8   18248755-18258723      + |          10
##     100    chr20   43248163-43280376      - |         100
##    1000    chr18   25530930-25757445      - |        1000
##   10000     chr1 243651535-244006886      - |       10000
##     ...      ...                 ...    ... .         ...
##    9991     chr9 114979995-115095944      - |        9991
##    9992    chr21   35736323-35743440      + |        9992
##    9993    chr22   19023795-19109967      - |        9993
##    9994     chr6   90539619-90584155      + |        9994
##    9997    chr22   50961997-50964905      - |        9997
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
</div>
<div id="grangeslist" class="section level2">
<h2 class="hasAnchor">
<a href="#grangeslist" class="anchor"></a>GRangesList</h2>
<p>Or the genomic position of all transcripts <strong>by</strong> gene:</p>
<div class="sourceCode" id="cb45"><html><body><pre class="r"><span class="no">txs</span> <span class="kw">&lt;-</span> <span class="fu">transcriptsBy</span>(<span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>)
<span class="no">txs</span></pre></body></html></div>
<pre><code>## GRangesList object of length 23459:
## $`1`
## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames            ranges strand |     tx_id     tx_name
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]    chr19 58858172-58864865      - |     70455  uc002qsd.4
##   [2]    chr19 58859832-58874214      - |     70456  uc002qsf.2
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
## 
## $`10`
## GRanges object with 1 range and 2 metadata columns:
##       seqnames            ranges strand |     tx_id     tx_name
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]     chr8 18248755-18258723      + |     31944  uc003wyw.1
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
## 
## $`100`
## GRanges object with 1 range and 2 metadata columns:
##       seqnames            ranges strand |     tx_id     tx_name
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]    chr20 43248163-43280376      - |     72132  uc002xmj.3
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
## 
## ...
## &lt;23456 more elements&gt;</code></pre>
</div>
<div id="other-accessors" class="section level2">
<h2 class="hasAnchor">
<a href="#other-accessors" class="anchor"></a>Other accessors</h2>
<ul>
<li><p>Positional information can be extracted for <code>transcripts</code>, <code>genes</code>, coding sequences (<code>cds</code>), <code>promoters</code> and <code>exons</code>.</p></li>
<li><p>Positional information can be extracted for most of the above, grouped by a second element. For example, our <code>transcriptsBy</code> call was all transcripts, grouped by gene.</p></li>
<li><p>More detail on these *Ranges objects is beyond the scope of this workshop, but why we want them is not.</p></li>
</ul>
</div>
<div id="why-ranges-objects" class="section level2">
<h2 class="hasAnchor">
<a href="#why-ranges-objects" class="anchor"></a>Why *Ranges objects</h2>
<p>The main rationale for *Ranges objects is to allow us to easily select and subset data based on genomic position information. This is really powerful!</p>
<p><code>GRanges</code> and <code>GRangesLists</code> act like data.frames and lists, and can be subsetted using the <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code> function. As a really artificial example:</p>
<div class="sourceCode" id="cb47"><html><body><pre class="r"><span class="no">txs</span>[<span class="no">txs</span> <span class="kw">%over%</span> <span class="no">gns</span>[<span class="fl">1</span>:<span class="fl">2</span>,]]</pre></body></html></div>
<pre><code>## GRangesList object of length 3:
## $`1`
## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames            ranges strand |     tx_id     tx_name
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]    chr19 58858172-58864865      - |     70455  uc002qsd.4
##   [2]    chr19 58859832-58874214      - |     70456  uc002qsf.2
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
## 
## $`10`
## GRanges object with 1 range and 2 metadata columns:
##       seqnames            ranges strand |     tx_id     tx_name
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]     chr8 18248755-18258723      + |     31944  uc003wyw.1
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
## 
## $`162968`
## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames            ranges strand |     tx_id     tx_name
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]    chr19 58865723-58874214      - |     70457  uc002qsh.2
##   [2]    chr19 58865723-58874214      - |     70458  uc002qsi.2
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
</div>
<div id="ranges-use-cases" class="section level2">
<h2 class="hasAnchor">
<a href="#ranges-use-cases" class="anchor"></a>*Ranges use cases</h2>
<ul>
<li><p>Gene expression changes near differentially methylated CpG islands</p></li>
<li><p>Closest genes to a set of interesting SNPs</p></li>
<li><p>Genes near DNAseI hypersensitivity clusters</p></li>
<li><p>Number of CpGs measured over Gene X by Chip Y</p></li>
</ul>
</div>
<div id="summarizedexperiment-objects" class="section level2">
<h2 class="hasAnchor">
<a href="#summarizedexperiment-objects" class="anchor"></a>SummarizedExperiment objects</h2>
<p>SummarizedExperiment objects are like ExpressionSets, but the row-wise annotations are GRanges, so you can subset by genomic locations:</p>
<p><img src="subset.png" width="65%"></p>
<p>SummarizedExperiment objects are popular objects for representing expression data and other rectangular data (feature x sample data). Incoming packages are now strongly recommended to use this class representation instead of ExpressionSet.</p>
</div>
<div id="txdb-exercises" class="section level2">
<h2 class="hasAnchor">
<a href="#txdb-exercises" class="anchor"></a>TxDb exercises</h2>
<ul>
<li><p>How many transcripts does PPARG have, according to UCSC?</p></li>
<li><p>Does Ensembl agree?</p></li>
<li>
<p>How many genes are between 2858473 and 3271812 on chr2 in the hg19 genome?</p>
<ul>
<li>Hint: you make a <code>GRanges</code> like this - <code>GRanges("chr2", IRanges(2858473,3271812))</code>
</li>
</ul>
</li>
</ul>
</div>
<div id="organismdb-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#organismdb-packages" class="anchor"></a>OrganismDb packages</h2>
<p>OrganismDb packages are meta-packages that contain an OrgDb, a TxDb, and a GO.db package and allow cross-queries between those packages.</p>
<p>All previous accessors work; <code>select</code>, <code>mapIds</code>, <code>transcripts</code>, etc.</p>
<div class="sourceCode" id="cb49"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">Homo.sapiens</span>)
<span class="no">Homo.sapiens</span></pre></body></html></div>
<pre><code>## OrganismDb Object:
## # Includes GODb Object:  GO.db 
## # With data about:  Gene Ontology 
## # Includes OrgDb Object:  org.Hs.eg.db 
## # Gene data about:  Homo sapiens 
## # Taxonomy Id:  9606 
## # Includes TxDb Object:  TxDb.Hsapiens.UCSC.hg19.knownGene 
## # Transcriptome data about:  Homo sapiens 
## # Based on genome:  hg19 
## # The OrgDb gene id ENTREZID is mapped to the TxDb gene id GENEID .</code></pre>
</div>
<div id="organismdb-packages-1" class="section level2">
<h2 class="hasAnchor">
<a href="#organismdb-packages-1" class="anchor"></a>OrganismDb packages</h2>
<ul>
<li><p>Updateable - can change TxDb object</p></li>
<li><p>columns and keytypes span all underlying objects</p></li>
<li><p>Calls to TxDb accessors include a ‘columns’ argument</p></li>
</ul>
<div class="sourceCode" id="cb51"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu">genes</span>(<span class="no">Homo.sapiens</span>, <span class="kw">columns</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"ENTREZID"</span>,<span class="st">"ALIAS"</span>,<span class="st">"UNIPROT"</span>)),<span class="fl">4</span>)</pre></body></html></div>
<pre><code>##   403 genes were dropped because they have exons located on both strands
##   of the same reference sequence or on more than one reference sequence,
##   so cannot be represented by a single genomic range.
##   Use 'single.strand.genes.only=FALSE' to get all the genes in a
##   GRangesList object, or use suppressMessages() to suppress this message.</code></pre>
<pre><code>## 'select()' returned 1:many mapping between keys and columns</code></pre>
<pre><code>## GRanges object with 4 ranges and 3 metadata columns:
##        seqnames            ranges strand |                 ALIAS
##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |       &lt;CharacterList&gt;
##      1    chr19 58858172-58874214      - |       A1B,ABG,GAB,...
##     10     chr8 18248755-18258723      + |   AAC2,NAT-2,PNAT,...
##    100    chr20 43248163-43280376      - |                   ADA
##   1000    chr18 25530930-25757445      - | CD325,CDHN,CDw325,...
##                         UNIPROT        ENTREZID
##                 &lt;CharacterList&gt; &lt;CharacterList&gt;
##      1            P04217,V9HWD8               1
##     10            A4Z6T7,P11245              10
##    100 A0A0S2Z381,P00813,F5GWI4             100
##   1000        P19022,A0A024RC42            1000
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
</div>
<div id="organismdb-exercises" class="section level2">
<h2 class="hasAnchor">
<a href="#organismdb-exercises" class="anchor"></a>OrganismDb exercises</h2>
<ul>
<li><p>Get all the GO terms for BRCA1</p></li>
<li><p>What gene does the UCSC transcript ID uc002fai.3 map to?</p></li>
<li><p>How many other transcripts does that gene have?</p></li>
<li><p>Get all the transcripts from the hg19 genome build, along with their Ensembl gene ID, UCSC transcript ID and gene symbol</p></li>
</ul>
</div>
<div id="organism-dplyr-package" class="section level2">
<h2 class="hasAnchor">
<a href="#organism-dplyr-package" class="anchor"></a>Organism.dplyr package</h2>
<ul>
<li><p>Combines the data from TxDb and Org.Db associated packages into local database.</p></li>
<li>
<p>Allows functions from both <em>org.*</em> and <em>TxDb.*</em></p>
<ul>
<li>
<code>keytypes()</code>, <code>select()</code>, …</li>
<li>
<code><a href="https://rdrr.io/pkg/Organism.dplyr/man/Genomic-Extractors.html">exons()</a></code>, <code><a href="https://rdrr.io/pkg/Organism.dplyr/man/Genomic-Extractors.html">promoters()</a></code>, …</li>
</ul>
</li>
<li><p>Allows for filtering and display of combined TxDb and Org.Db information through <code>dplyr</code> functions.</p></li>
</ul>
<div class="sourceCode" id="cb55"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">Organism.dplyr</span>)

<span class="co"># src = src_organism("TxDb.Hsapiens.UCSC.hg19.knownGene")</span>
<span class="no">src</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Organism.dplyr/man/src.html">src_organism</a></span>(<span class="kw">dbpath</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Organism.dplyr/man/utils.html">hg38light</a></span>())
<span class="no">src</span></pre></body></html></div>
<pre><code>## src:  sqlite 3.30.1 [/usr/local/lib/R/site-library/Organism.dplyr/extdata/light.hg38.knownGene.sqlite]
## tbls: id, id_accession, id_go, id_go_all, id_omim_pm, id_protein,
##   id_transcript, ranges_cds, ranges_exon, ranges_gene, ranges_tx</code></pre>
</div>
<div id="organism-dplyr" class="section level2">
<h2 class="hasAnchor">
<a href="#organism-dplyr" class="anchor"></a>Organism.dplyr</h2>
<p>Get promoters from a TxDb object (we use a small version)</p>
<div class="sourceCode" id="cb57"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(<span class="kw">width</span> <span class="kw">=</span> <span class="fl">120</span>)
<span class="fu"><a href="https://rdrr.io/pkg/Organism.dplyr/man/Genomic-Extractors.html">promoters</a></span>(<span class="no">src</span>)</pre></body></html></div>
<pre><code>## GRanges object with 96 ranges and 2 metadata columns:
##                                  seqnames              ranges strand |     tx_id            tx_name
##                                     &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;        &lt;character&gt;
##    ENST00000336199.9                 chr1 243843037-243845236      - |     19678  ENST00000336199.9
##    ENST00000366540.5                 chr1 243843385-243845584      - |     19679  ENST00000366540.5
##   ENST00000263826.11                 chr1 243843083-243845282      - |     19680 ENST00000263826.11
##    ENST00000673466.1                 chr1 243850044-243852243      - |     19681  ENST00000673466.1
##    ENST00000672460.1                 chr1 243664613-243666812      - |     19682  ENST00000672460.1
##                  ...                  ...                 ...    ... .       ...                ...
##    ENST00000646420.1 chr11_KZ559110v1_alt         38835-41034      + |    240550  ENST00000646420.1
##    ENST00000642929.1 chr11_KZ559110v1_alt         38852-41051      + |    240551  ENST00000642929.1
##    ENST00000644696.1 chr11_KZ559110v1_alt         38861-41060      + |    240552  ENST00000644696.1
##    ENST00000646644.1 chr11_KZ559110v1_alt         39187-41386      + |    240553  ENST00000646644.1
##    ENST00000645918.1 chr11_KZ559110v1_alt         66615-68814      + |    240554  ENST00000645918.1
##   -------
##   seqinfo: 595 sequences (1 circular) from hg38 genome</code></pre>
</div>
<div id="organism-dplyr-1" class="section level2">
<h2 class="hasAnchor">
<a href="#organism-dplyr-1" class="anchor"></a>Organism.dplyr</h2>
<p>Extract a table from the underlying database</p>
<div class="sourceCode" id="cb59"><html><body><pre class="r"><span class="fu">tbl</span>(<span class="no">src</span>, <span class="st">"id"</span>)</pre></body></html></div>
<pre><code>## # Source:   table&lt;id&gt; [?? x 6]
## # Database: sqlite 3.30.1 []
##    entrez map      ensembl         symbol genename               alias   
##    &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;           &lt;chr&gt;  &lt;chr&gt;                  &lt;chr&gt;   
##  1 1      19q13.43 ENSG00000121410 A1BG   alpha-1-B glycoprotein A1B     
##  2 1      19q13.43 ENSG00000121410 A1BG   alpha-1-B glycoprotein ABG     
##  3 1      19q13.43 ENSG00000121410 A1BG   alpha-1-B glycoprotein GAB     
##  4 1      19q13.43 ENSG00000121410 A1BG   alpha-1-B glycoprotein HYST2477
##  5 1      19q13.43 ENSG00000121410 A1BG   alpha-1-B glycoprotein A1BG    
##  6 10     8p22     ENSG00000156006 NAT2   N-acetyltransferase 2  AAC2    
##  7 10     8p22     ENSG00000156006 NAT2   N-acetyltransferase 2  NAT-2   
##  8 10     8p22     ENSG00000156006 NAT2   N-acetyltransferase 2  PNAT    
##  9 10     8p22     ENSG00000156006 NAT2   N-acetyltransferase 2  NAT2    
## 10 100    20q13.12 ENSG00000196839 ADA    adenosine deaminase    ADA     
## # … with more rows</code></pre>
</div>
<div id="organism-dplyr-2" class="section level2">
<h2 class="hasAnchor">
<a href="#organism-dplyr-2" class="anchor"></a>Organism.dplyr</h2>
<p>Make a complex query between tables in the underlying database</p>
<div class="sourceCode" id="cb61"><html><body><pre class="r"><span class="fu">inner_join</span>(<span class="fu">tbl</span>(<span class="no">src</span>, <span class="st">"id"</span>), <span class="fu">tbl</span>(<span class="no">src</span>, <span class="st">"ranges_gene"</span>)) <span class="kw">%&gt;%</span>
            <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(<span class="no">symbol</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"ADA"</span>, <span class="st">"NAT2"</span>)) <span class="kw">%&gt;%</span>
            <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="no">gene_chrom</span>, <span class="no">gene_start</span>, <span class="no">gene_end</span>,
            <span class="no">gene_strand</span>, <span class="no">symbol</span>, <span class="no">alias</span>, <span class="no">map</span>)</pre></body></html></div>
<pre><code>## Joining, by = "entrez"</code></pre>
<pre><code>## # Source:   lazy query [?? x 7]
## # Database: sqlite 3.30.1 []
##   gene_chrom gene_start gene_end gene_strand symbol alias map     
##   &lt;chr&gt;           &lt;int&gt;    &lt;int&gt; &lt;chr&gt;       &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;   
## 1 chr8         18391282 18401218 +           NAT2   AAC2  8p22    
## 2 chr8         18391282 18401218 +           NAT2   NAT-2 8p22    
## 3 chr8         18391282 18401218 +           NAT2   PNAT  8p22    
## 4 chr8         18391282 18401218 +           NAT2   NAT2  8p22    
## 5 chr20        44619522 44652233 -           ADA    ADA   20q13.12</code></pre>
</div>
<div id="organism-dplyr-exercises" class="section level2">
<h2 class="hasAnchor">
<a href="#organism-dplyr-exercises" class="anchor"></a>Organism.dplyr exercises</h2>
<ul>
<li><p>How many supported organisms are implemented in Organism.dplyr?</p></li>
<li><p>Display the ensembl Id and genename description for symbol “NAT2”.</p></li>
<li><p>Show all the alias for “NAT2” in the database.</p></li>
<li><p>Display Gene ontology (GO) information for gene symbol “Nat2”.</p></li>
</ul>
</div>
<div id="bsgenome-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#bsgenome-packages" class="anchor"></a>BSgenome packages</h2>
<p>BSgenome packages contain sequence information for a given species/build. There are many such packages - you can get a listing using <code>available.genomes</code></p>
<div class="sourceCode" id="cb64"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">BSgenome</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/BSgenome/man/available.genomes.html">available.genomes</a></span>())</pre></body></html></div>
<pre><code>## [1] "BSgenome.Alyrata.JGI.v1"                 "BSgenome.Amellifera.BeeBase.assembly4"  
## [3] "BSgenome.Amellifera.UCSC.apiMel2"        "BSgenome.Amellifera.UCSC.apiMel2.masked"
## [5] "BSgenome.Aofficinalis.NCBI.V1"           "BSgenome.Athaliana.TAIR.04232008"</code></pre>
</div>
<div id="bsgenome-packages-1" class="section level2">
<h2 class="hasAnchor">
<a href="#bsgenome-packages-1" class="anchor"></a>BSgenome packages</h2>
<p>We can load and inspect a BSgenome package</p>
<div class="sourceCode" id="cb66"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">BSgenome.Hsapiens.UCSC.hg19</span>)
<span class="no">Hsapiens</span></pre></body></html></div>
<pre><code>## Human genome:
## # organism: Homo sapiens (Human)
## # provider: UCSC
## # provider version: hg19
## # release date: June 2013
## # release name: Genome Reference Consortium GRCh37.p13
## # 298 sequences:
## #   chr1                  chr2                  chr3                  chr4                  chr5                 
## #   chr6                  chr7                  chr8                  chr9                  chr10                
## #   chr11                 chr12                 chr13                 chr14                 chr15                
## #   chr16                 chr17                 chr18                 chr19                 chr20                
## #   chr21                 chr22                 chrX                  chrY                  chrM                 
## #   ...                   ...                   ...                   ...                   ...                  
## #   chr18_gl383570_alt    chr18_gl383571_alt    chr18_gl383572_alt    chr19_gl383573_alt    chr19_gl383574_alt   
## #   chr19_gl383575_alt    chr19_gl383576_alt    chr19_gl949746_alt    chr19_gl949747_alt    chr19_gl949748_alt   
## #   chr19_gl949749_alt    chr19_gl949750_alt    chr19_gl949751_alt    chr19_gl949752_alt    chr19_gl949753_alt   
## #   chr20_gl383577_alt    chr21_gl383578_alt    chr21_gl383579_alt    chr21_gl383580_alt    chr21_gl383581_alt   
## #   chr22_gl383582_alt    chr22_gl383583_alt    chr22_kb663609_alt                                               
## # (use 'seqnames()' to see all the sequence names, use the '$' or '[[' operator to access a given sequence)</code></pre>
</div>
<div id="bsgenome-packages-2" class="section level2">
<h2 class="hasAnchor">
<a href="#bsgenome-packages-2" class="anchor"></a>BSgenome packages</h2>
<p>The main accessor is <code>getSeq</code>, and you can get data by sequence (e.g., entire chromosome or unplaced scaffold), or by passing in a GRanges object, to get just a region.</p>
<div class="sourceCode" id="cb68"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/BSgenome/man/getSeq-methods.html">getSeq</a></span>(<span class="no">Hsapiens</span>, <span class="st">"chr1"</span>)</pre></body></html></div>
<pre><code>## 249250621-letter DNAString object
## seq: NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN</code></pre>
<div class="sourceCode" id="cb70"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/BSgenome/man/getSeq-methods.html">getSeq</a></span>(<span class="no">Hsapiens</span>, <span class="no">gns</span>[<span class="st">"5467"</span>,])</pre></body></html></div>
<pre><code>## DNAStringSet object of length 1:
##     width seq                                                                                       names               
## [1] 85634 GCGGAGCGTGTGACGCTGCGGCCGCCGCGGACCTGGGGATTAA...ACTTTAAATAAATGGGAATTAAATATTTAAGAGCTGACTGGAA 5467</code></pre>
<p>The Biostrings package contains most of the code for dealing with these <code>*StringSet</code> objects - please see the Biostrings vignettes and help pages for more information.</p>
</div>
<div id="bsgenome-exercises" class="section level2">
<h2 class="hasAnchor">
<a href="#bsgenome-exercises" class="anchor"></a>BSgenome exercises</h2>
<ul>
<li>Get the sequences for all transcripts of the TP53 gene</li>
</ul>
</div>
<div id="annotationhub" class="section level2">
<h2 class="hasAnchor">
<a href="#annotationhub" class="anchor"></a>AnnotationHub</h2>
<p>AnnotationHub is a package that allows us to query and download many different annotation objects, without having to explicitly install them.</p>
<div class="sourceCode" id="cb72"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">AnnotationHub</span>)
<span class="no">hub</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html">AnnotationHub</a></span>()</pre></body></html></div>
<pre><code>## snapshotDate(): 2020-06-18</code></pre>
<div class="sourceCode" id="cb74"><html><body><pre class="r"><span class="no">hub</span></pre></body></html></div>
<pre><code>## AnnotationHub with 50279 records
## # snapshotDate(): 2020-06-18
## # $dataprovider: BroadInstitute, Ensembl, UCSC, ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/, Haemcode, FungiDB, Inparanoid...
## # $species: Homo sapiens, Mus musculus, Drosophila melanogaster, Bos taurus, Pan troglodytes, Rattus norvegicus, Dan...
## # $rdataclass: GRanges, BigWigFile, TwoBitFile, Rle, EnsDb, OrgDb, ChainFile, TxDb, Inparanoid8Db, data.frame
## # additional mcols(): taxonomyid, genome, description, coordinate_1_based, maintainer, rdatadateadded,
## #   preparerclass, tags, rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g., 'object[["AH5012"]]' 
## 
##             title                                                                  
##   AH5012  | Chromosome Band                                                        
##   AH5013  | STS Markers                                                            
##   AH5014  | FISH Clones                                                            
##   AH5015  | Recomb Rate                                                            
##   AH5016  | ENCODE Pilot                                                           
##   ...       ...                                                                    
##   AH83111 | Zosterops_lateralis_melanops.ASM128173v1.cdna.all.2bit                 
##   AH83112 | Zosterops_lateralis_melanops.ASM128173v1.dna_rm.toplevel.2bit          
##   AH83113 | Zosterops_lateralis_melanops.ASM128173v1.dna_sm.toplevel.2bit          
##   AH83114 | Zosterops_lateralis_melanops.ASM128173v1.ncrna.2bit                    
##   AH83115 | Mapping table of metabolite IDs and common names from different sources</code></pre>
</div>
<div id="querying-annotationhub" class="section level2">
<h2 class="hasAnchor">
<a href="#querying-annotationhub" class="anchor"></a>Querying AnnotationHub</h2>
<p>Finding the ‘right’ resource on AnnotationHub is like using Google - a well posed query is necessary to find what you are after. Useful queries are based on</p>
<ul>
<li><p>Data provider</p></li>
<li><p>Data class</p></li>
<li><p>Species</p></li>
<li><p>Data source</p></li>
</ul>
<div class="sourceCode" id="cb76"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="fu">mcols</span>(<span class="no">hub</span>))</pre></body></html></div>
<pre><code>##  [1] "title"              "dataprovider"       "species"            "taxonomyid"         "genome"            
##  [6] "description"        "coordinate_1_based" "maintainer"         "rdatadateadded"     "preparerclass"     
## [11] "tags"               "rdataclass"         "rdatapath"          "sourceurl"          "sourcetype"</code></pre>
</div>
<div id="annotationhub-data-providers" class="section level2">
<h2 class="hasAnchor">
<a href="#annotationhub-data-providers" class="anchor"></a>AnnotationHub Data providers</h2>
<div class="sourceCode" id="cb78"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">hub</span>$<span class="no">dataprovider</span>)</pre></body></html></div>
<pre><code>##  [1] "UCSC"                                                "Ensembl"                                            
##  [3] "RefNet"                                              "Inparanoid8"                                        
##  [5] "NHLBI"                                               "ChEA"                                               
##  [7] "Pazar"                                               "NIH Pathway Interaction Database"                   
##  [9] "Haemcode"                                            "BroadInstitute"                                     
## [11] "PRIDE"                                               "Gencode"                                            
## [13] "CRIBI"                                               "Genoscope"                                          
## [15] "MISO, VAST-TOOLS, UCSC"                              "UWashington"                                        
## [17] "Stanford"                                            "dbSNP"                                              
## [19] "BioMart"                                             "GeneOntology"                                       
## [21] "KEGG"                                                "URGI"                                               
## [23] "EMBL-EBI"                                            "MicrosporidiaDB"                                    
## [25] "FungiDB"                                             "TriTrypDB"                                          
## [27] "ToxoDB"                                              "AmoebaDB"                                           
## [29] "PlasmoDB"                                            "PiroplasmaDB"                                       
## [31] "CryptoDB"                                            "TrichDB"                                            
## [33] "GiardiaDB"                                           "The Gene Ontology Consortium"                       
## [35] "ENCODE Project"                                      "SchistoDB"                                          
## [37] "NCBI/UniProt"                                        "GENCODE"                                            
## [39] "http://www.pantherdb.org"                            "RMBase v2.0"                                        
## [41] "snoRNAdb"                                            "tRNAdb"                                             
## [43] "NCBI"                                                "DrugAge, DrugBank, Broad Institute, Broad Institute"
## [45] "DrugAge"                                             "DrugBank"                                           
## [47] "Broad Institute"                                     "ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/"              
## [49] "HMDB, EMBL-EBI, EPA"</code></pre>
</div>
<div id="annotationhub-data-classes" class="section level2">
<h2 class="hasAnchor">
<a href="#annotationhub-data-classes" class="anchor"></a>AnnotationHub Data classes</h2>
<div class="sourceCode" id="cb80"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">hub</span>$<span class="no">rdataclass</span>)</pre></body></html></div>
<pre><code>##  [1] "GRanges"                           "data.frame"                        "Inparanoid8Db"                    
##  [4] "TwoBitFile"                        "ChainFile"                         "SQLiteConnection"                 
##  [7] "biopax"                            "BigWigFile"                        "AAStringSet"                      
## [10] "MSnSet"                            "mzRpwiz"                           "mzRident"                         
## [13] "list"                              "TxDb"                              "Rle"                              
## [16] "EnsDb"                             "VcfFile"                           "igraph"                           
## [19] "data.frame, DNAStringSet, GRanges" "sqlite"                            "data.table"                       
## [22] "character"                         "SQLite"                            "SQLiteFile"                       
## [25] "OrgDb"                             "Tibble"</code></pre>
</div>
<div id="annotationhub-species" class="section level2">
<h2 class="hasAnchor">
<a href="#annotationhub-species" class="anchor"></a>AnnotationHub Species</h2>
<div class="sourceCode" id="cb82"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">hub</span>$<span class="no">species</span>))</pre></body></html></div>
<pre><code>## [1] "Homo sapiens"         "Vicugna pacos"        "Dasypus novemcinctus" "Otolemur garnettii"   "Papio hamadryas"     
## [6] "Papio anubis"</code></pre>
<div class="sourceCode" id="cb84"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">hub</span>$<span class="no">species</span>))</pre></body></html></div>
<pre><code>## [1] 2116</code></pre>
</div>
<div id="annotationhub-data-sources" class="section level2">
<h2 class="hasAnchor">
<a href="#annotationhub-data-sources" class="anchor"></a>AnnotationHub Data sources</h2>
<div class="sourceCode" id="cb86"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">hub</span>$<span class="no">sourcetype</span>)</pre></body></html></div>
<pre><code>##  [1] "UCSC track"   "GTF"          "TSV"          "Inparanoid"   "TwoBit"       "Chain"        "GRASP"       
##  [8] "Zip"          "CSV"          "BioPax"       "BioPaxLevel2" "RData"        "BED"          "BigWig"      
## [15] "tab"          "FASTA"        "mzTab"        "mzML"         "mzid"         "GFF"          "ensembl"     
## [22] "VCF"          "TXT"          "JSON"         "XML"          "XLS/XLSX"     "NCBI/ensembl" "NCBI/UniProt"</code></pre>
</div>
<div id="annotationhub-query" class="section level2">
<h2 class="hasAnchor">
<a href="#annotationhub-query" class="anchor"></a>AnnotationHub query</h2>
<div class="sourceCode" id="cb88"><html><body><pre class="r"><span class="no">qry</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html">query</a></span>(<span class="no">hub</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"granges"</span>,<span class="st">"homo sapiens"</span>,<span class="st">"ensembl"</span>))
<span class="no">qry</span></pre></body></html></div>
<pre><code>## AnnotationHub with 84 records
## # snapshotDate(): 2020-06-18
## # $dataprovider: Ensembl, UCSC
## # $species: Homo sapiens
## # $rdataclass: GRanges
## # additional mcols(): taxonomyid, genome, description, coordinate_1_based, maintainer, rdatadateadded,
## #   preparerclass, tags, rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g., 'object[["AH5046"]]' 
## 
##             title                                           
##   AH5046  | Ensembl Genes                                   
##   AH5160  | Ensembl Genes                                   
##   AH5311  | Ensembl Genes                                   
##   AH5434  | Ensembl Genes                                   
##   AH5435  | Ensembl EST Genes                               
##   ...       ...                                             
##   AH79161 | Homo_sapiens.GRCh38.99.gtf                      
##   AH80074 | Homo_sapiens.GRCh38.100.abinitio.gtf            
##   AH80075 | Homo_sapiens.GRCh38.100.chr.gtf                 
##   AH80076 | Homo_sapiens.GRCh38.100.chr_patch_hapl_scaff.gtf
##   AH80077 | Homo_sapiens.GRCh38.100.gtf</code></pre>
</div>
<div id="annotationhub-query-1" class="section level2">
<h2 class="hasAnchor">
<a href="#annotationhub-query-1" class="anchor"></a>AnnotationHub query</h2>
<div class="sourceCode" id="cb90"><html><body><pre class="r"><span class="no">qry</span>$<span class="no">sourceurl</span></pre></body></html></div>
<pre><code>##  [1] "rtracklayer://hgdownload.cse.ucsc.edu/goldenpath/hg19/database/ensGene"                                    
##  [2] "rtracklayer://hgdownload.cse.ucsc.edu/goldenpath/hg18/database/ensGene"                                    
##  [3] "rtracklayer://hgdownload.cse.ucsc.edu/goldenpath/hg17/database/ensGene"                                    
##  [4] "rtracklayer://hgdownload.cse.ucsc.edu/goldenpath/hg16/database/ensGene"                                    
##  [5] "rtracklayer://hgdownload.cse.ucsc.edu/goldenpath/hg16/database/ensEstGene"                                 
##  [6] "ftp://ftp.ensembl.org/pub/release-70/gtf/homo_sapiens/Homo_sapiens.GRCh37.70.gtf.gz"                       
##  [7] "ftp://ftp.ensembl.org/pub/release-69/gtf/homo_sapiens/Homo_sapiens.GRCh37.69.gtf.gz"                       
##  [8] "ftp://ftp.ensembl.org/pub/release-71/gtf/homo_sapiens/Homo_sapiens.GRCh37.71.gtf.gz"                       
##  [9] "ftp://ftp.ensembl.org/pub/release-72/gtf/homo_sapiens/Homo_sapiens.GRCh37.72.gtf.gz"                       
## [10] "ftp://ftp.ensembl.org/pub/release-73/gtf/homo_sapiens/Homo_sapiens.GRCh37.73.gtf.gz"                       
## [11] "ftp://ftp.ensembl.org/pub/release-74/gtf/homo_sapiens/Homo_sapiens.GRCh37.74.gtf.gz"                       
## [12] "ftp://ftp.ensembl.org/pub/release-75/gtf/homo_sapiens/Homo_sapiens.GRCh37.75.gtf.gz"                       
## [13] "ftp://ftp.ensembl.org/pub/release-78/gtf/homo_sapiens/Homo_sapiens.GRCh38.78.gtf.gz"                       
## [14] "ftp://ftp.ensembl.org/pub/release-76/gtf/homo_sapiens/Homo_sapiens.GRCh38.76.gtf.gz"                       
## [15] "ftp://ftp.ensembl.org/pub/release-79/gtf/homo_sapiens/Homo_sapiens.GRCh38.79.gtf.gz"                       
## [16] "ftp://ftp.ensembl.org/pub/release-77/gtf/homo_sapiens/Homo_sapiens.GRCh38.77.gtf.gz"                       
## [17] "ftp://ftp.ensembl.org/pub/release-80/gtf/homo_sapiens/Homo_sapiens.GRCh38.80.gtf.gz"                       
## [18] "ftp://ftp.ensembl.org/pub/release-81/gtf/homo_sapiens/Homo_sapiens.GRCh38.81.gtf.gz"                       
## [19] "ftp://ftp.ensembl.org/pub/release-82/gtf/homo_sapiens/Homo_sapiens.GRCh38.82.gtf.gz"                       
## [20] "ftp://ftp.ensembl.org/pub/release-83/gtf/homo_sapiens/Homo_sapiens.GRCh38.83.gtf.gz"                       
## [21] "ftp://ftp.ensembl.org/pub/release-84/gtf/homo_sapiens/Homo_sapiens.GRCh38.84.abinitio.gtf.gz"              
## [22] "ftp://ftp.ensembl.org/pub/release-84/gtf/homo_sapiens/Homo_sapiens.GRCh38.84.chr.gtf.gz"                   
## [23] "ftp://ftp.ensembl.org/pub/release-84/gtf/homo_sapiens/Homo_sapiens.GRCh38.84.chr_patch_hapl_scaff.gtf.gz"  
## [24] "ftp://ftp.ensembl.org/pub/release-84/gtf/homo_sapiens/Homo_sapiens.GRCh38.84.gtf.gz"                       
## [25] "ftp://ftp.ensembl.org/pub/release-85/gtf/homo_sapiens/Homo_sapiens.GRCh38.85.abinitio.gtf.gz"              
## [26] "ftp://ftp.ensembl.org/pub/release-85/gtf/homo_sapiens/Homo_sapiens.GRCh38.85.chr.gtf.gz"                   
## [27] "ftp://ftp.ensembl.org/pub/release-85/gtf/homo_sapiens/Homo_sapiens.GRCh38.85.chr_patch_hapl_scaff.gtf.gz"  
## [28] "ftp://ftp.ensembl.org/pub/release-85/gtf/homo_sapiens/Homo_sapiens.GRCh38.85.gtf.gz"                       
## [29] "ftp://ftp.ensembl.org/pub/release-86/gtf/homo_sapiens/Homo_sapiens.GRCh38.86.abinitio.gtf.gz"              
## [30] "ftp://ftp.ensembl.org/pub/release-86/gtf/homo_sapiens/Homo_sapiens.GRCh38.86.chr.gtf.gz"                   
## [31] "ftp://ftp.ensembl.org/pub/release-86/gtf/homo_sapiens/Homo_sapiens.GRCh38.86.chr_patch_hapl_scaff.gtf.gz"  
## [32] "ftp://ftp.ensembl.org/pub/release-86/gtf/homo_sapiens/Homo_sapiens.GRCh38.86.gtf.gz"                       
## [33] "ftp://ftp.ensembl.org/pub/release-87/gtf/homo_sapiens/Homo_sapiens.GRCh38.87.abinitio.gtf.gz"              
## [34] "ftp://ftp.ensembl.org/pub/release-87/gtf/homo_sapiens/Homo_sapiens.GRCh38.87.chr.gtf.gz"                   
## [35] "ftp://ftp.ensembl.org/pub/release-87/gtf/homo_sapiens/Homo_sapiens.GRCh38.87.chr_patch_hapl_scaff.gtf.gz"  
## [36] "ftp://ftp.ensembl.org/pub/release-87/gtf/homo_sapiens/Homo_sapiens.GRCh38.87.gtf.gz"                       
## [37] "ftp://ftp.ensembl.org/pub/release-88/gtf/homo_sapiens/Homo_sapiens.GRCh38.88.abinitio.gtf.gz"              
## [38] "ftp://ftp.ensembl.org/pub/release-88/gtf/homo_sapiens/Homo_sapiens.GRCh38.88.chr.gtf.gz"                   
## [39] "ftp://ftp.ensembl.org/pub/release-88/gtf/homo_sapiens/Homo_sapiens.GRCh38.88.chr_patch_hapl_scaff.gtf.gz"  
## [40] "ftp://ftp.ensembl.org/pub/release-88/gtf/homo_sapiens/Homo_sapiens.GRCh38.88.gtf.gz"                       
## [41] "ftp://ftp.ensembl.org/pub/release-89/gtf/homo_sapiens/Homo_sapiens.GRCh38.89.abinitio.gtf.gz"              
## [42] "ftp://ftp.ensembl.org/pub/release-89/gtf/homo_sapiens/Homo_sapiens.GRCh38.89.chr.gtf.gz"                   
## [43] "ftp://ftp.ensembl.org/pub/release-89/gtf/homo_sapiens/Homo_sapiens.GRCh38.89.chr_patch_hapl_scaff.gtf.gz"  
## [44] "ftp://ftp.ensembl.org/pub/release-89/gtf/homo_sapiens/Homo_sapiens.GRCh38.89.gtf.gz"                       
## [45] "ftp://ftp.ensembl.org/pub/release-90/gtf/homo_sapiens/Homo_sapiens.GRCh38.90.abinitio.gtf.gz"              
## [46] "ftp://ftp.ensembl.org/pub/release-90/gtf/homo_sapiens/Homo_sapiens.GRCh38.90.chr.gtf.gz"                   
## [47] "ftp://ftp.ensembl.org/pub/release-90/gtf/homo_sapiens/Homo_sapiens.GRCh38.90.chr_patch_hapl_scaff.gtf.gz"  
## [48] "ftp://ftp.ensembl.org/pub/release-90/gtf/homo_sapiens/Homo_sapiens.GRCh38.90.gtf.gz"                       
## [49] "ftp://ftp.ensembl.org/pub/release-91/gtf/homo_sapiens/Homo_sapiens.GRCh38.91.abinitio.gtf.gz"              
## [50] "ftp://ftp.ensembl.org/pub/release-91/gtf/homo_sapiens/Homo_sapiens.GRCh38.91.chr.gtf.gz"                   
## [51] "ftp://ftp.ensembl.org/pub/release-91/gtf/homo_sapiens/Homo_sapiens.GRCh38.91.chr_patch_hapl_scaff.gtf.gz"  
## [52] "ftp://ftp.ensembl.org/pub/release-91/gtf/homo_sapiens/Homo_sapiens.GRCh38.91.gtf.gz"                       
## [53] "ftp://ftp.ensembl.org/pub/release-92/gtf/homo_sapiens/Homo_sapiens.GRCh38.92.abinitio.gtf.gz"              
## [54] "ftp://ftp.ensembl.org/pub/release-92/gtf/homo_sapiens/Homo_sapiens.GRCh38.92.chr.gtf.gz"                   
## [55] "ftp://ftp.ensembl.org/pub/release-92/gtf/homo_sapiens/Homo_sapiens.GRCh38.92.chr_patch_hapl_scaff.gtf.gz"  
## [56] "ftp://ftp.ensembl.org/pub/release-92/gtf/homo_sapiens/Homo_sapiens.GRCh38.92.gtf.gz"                       
## [57] "ftp://ftp.ensembl.org/pub/release-93/gtf/homo_sapiens/Homo_sapiens.GRCh38.93.abinitio.gtf.gz"              
## [58] "ftp://ftp.ensembl.org/pub/release-93/gtf/homo_sapiens/Homo_sapiens.GRCh38.93.chr.gtf.gz"                   
## [59] "ftp://ftp.ensembl.org/pub/release-93/gtf/homo_sapiens/Homo_sapiens.GRCh38.93.chr_patch_hapl_scaff.gtf.gz"  
## [60] "ftp://ftp.ensembl.org/pub/release-93/gtf/homo_sapiens/Homo_sapiens.GRCh38.93.gtf.gz"                       
## [61] "ftp://ftp.ensembl.org/pub/release-94/gtf/homo_sapiens/Homo_sapiens.GRCh38.94.abinitio.gtf.gz"              
## [62] "ftp://ftp.ensembl.org/pub/release-94/gtf/homo_sapiens/Homo_sapiens.GRCh38.94.chr.gtf.gz"                   
## [63] "ftp://ftp.ensembl.org/pub/release-94/gtf/homo_sapiens/Homo_sapiens.GRCh38.94.chr_patch_hapl_scaff.gtf.gz"  
## [64] "ftp://ftp.ensembl.org/pub/release-94/gtf/homo_sapiens/Homo_sapiens.GRCh38.94.gtf.gz"                       
## [65] "ftp://ftp.ensembl.org/pub/release-95/gtf/homo_sapiens/Homo_sapiens.GRCh38.95.abinitio.gtf.gz"              
## [66] "ftp://ftp.ensembl.org/pub/release-95/gtf/homo_sapiens/Homo_sapiens.GRCh38.95.chr.gtf.gz"                   
## [67] "ftp://ftp.ensembl.org/pub/release-95/gtf/homo_sapiens/Homo_sapiens.GRCh38.95.chr_patch_hapl_scaff.gtf.gz"  
## [68] "ftp://ftp.ensembl.org/pub/release-95/gtf/homo_sapiens/Homo_sapiens.GRCh38.95.gtf.gz"                       
## [69] "ftp://ftp.ensembl.org/pub/release-96/gtf/homo_sapiens/Homo_sapiens.GRCh38.96.abinitio.gtf.gz"              
## [70] "ftp://ftp.ensembl.org/pub/release-96/gtf/homo_sapiens/Homo_sapiens.GRCh38.96.chr.gtf.gz"                   
## [71] "ftp://ftp.ensembl.org/pub/release-96/gtf/homo_sapiens/Homo_sapiens.GRCh38.96.chr_patch_hapl_scaff.gtf.gz"  
## [72] "ftp://ftp.ensembl.org/pub/release-96/gtf/homo_sapiens/Homo_sapiens.GRCh38.96.gtf.gz"                       
## [73] "ftp://ftp.ensembl.org/pub/release-98/gtf/homo_sapiens/Homo_sapiens.GRCh38.98.abinitio.gtf.gz"              
## [74] "ftp://ftp.ensembl.org/pub/release-98/gtf/homo_sapiens/Homo_sapiens.GRCh38.98.chr.gtf.gz"                   
## [75] "ftp://ftp.ensembl.org/pub/release-98/gtf/homo_sapiens/Homo_sapiens.GRCh38.98.chr_patch_hapl_scaff.gtf.gz"  
## [76] "ftp://ftp.ensembl.org/pub/release-98/gtf/homo_sapiens/Homo_sapiens.GRCh38.98.gtf.gz"                       
## [77] "ftp://ftp.ensembl.org/pub/release-99/gtf/homo_sapiens/Homo_sapiens.GRCh38.99.abinitio.gtf.gz"              
## [78] "ftp://ftp.ensembl.org/pub/release-99/gtf/homo_sapiens/Homo_sapiens.GRCh38.99.chr.gtf.gz"                   
## [79] "ftp://ftp.ensembl.org/pub/release-99/gtf/homo_sapiens/Homo_sapiens.GRCh38.99.chr_patch_hapl_scaff.gtf.gz"  
## [80] "ftp://ftp.ensembl.org/pub/release-99/gtf/homo_sapiens/Homo_sapiens.GRCh38.99.gtf.gz"                       
## [81] "ftp://ftp.ensembl.org/pub/release-100/gtf/homo_sapiens/Homo_sapiens.GRCh38.100.abinitio.gtf.gz"            
## [82] "ftp://ftp.ensembl.org/pub/release-100/gtf/homo_sapiens/Homo_sapiens.GRCh38.100.chr.gtf.gz"                 
## [83] "ftp://ftp.ensembl.org/pub/release-100/gtf/homo_sapiens/Homo_sapiens.GRCh38.100.chr_patch_hapl_scaff.gtf.gz"
## [84] "ftp://ftp.ensembl.org/pub/release-100/gtf/homo_sapiens/Homo_sapiens.GRCh38.100.gtf.gz"</code></pre>
</div>
<div id="selecting-annotationhub-resource" class="section level2">
<h2 class="hasAnchor">
<a href="#selecting-annotationhub-resource" class="anchor"></a>Selecting AnnotationHub resource</h2>
<div class="sourceCode" id="cb92"><html><body><pre class="r"><span class="no">whatIwant</span> <span class="kw">&lt;-</span> <span class="no">qry</span><span class="kw">[[</span><span class="st">"AH80077"</span>]]</pre></body></html></div>
<p>We can use these data as they are, or convert to a TxDb format:</p>
<div class="sourceCode" id="cb93"><html><body><pre class="r"><span class="no">GRCh38TxDb</span> <span class="kw">&lt;-</span> <span class="fu">makeTxDbFromGRanges</span>(<span class="no">whatIwant</span>)
<span class="no">GRCh38TxDb</span></pre></body></html></div>
<pre><code>## TxDb object:
## # Db type: TxDb
## # Supporting package: GenomicFeatures
## # Genome: GRCh38.p13
## # Nb of transcripts: 227887
## # Db created by: GenomicFeatures package from Bioconductor
## # Creation time: 2020-06-26 15:27:41 +0000 (Fri, 26 Jun 2020)
## # GenomicFeatures version at creation time: 1.41.0
## # RSQLite version at creation time: 2.2.0
## # DBSCHEMAVERSION: 1.2</code></pre>
</div>
<div id="annotationhub-exercises" class="section level2">
<h2 class="hasAnchor">
<a href="#annotationhub-exercises" class="anchor"></a>AnnotationHub exercises</h2>
<ul>
<li><p>How many resources are on AnnotationHub for Atlantic salmon (Salmo salar)?</p></li>
<li><p>Get the most recent Ensembl build for domesticated dog (Canis familiaris) and make a TxDb</p></li>
</ul>
</div>
<div id="biomart" class="section level2">
<h2 class="hasAnchor">
<a href="#biomart" class="anchor"></a>biomaRt</h2>
<p>The biomaRt package allows queries to an Ensembl Biomart server. We can see the choices of servers that we can use:</p>
<div class="sourceCode" id="cb95"><html><body><pre class="r">library(biomaRt)
listMarts(host = "useast.ensembl.org")

               biomart                version
1 ENSEMBL_MART_ENSEMBL      Ensembl Genes 100
2   ENSEMBL_MART_MOUSE      Mouse strains 100
3     ENSEMBL_MART_SNP  Ensembl Variation 100
4 ENSEMBL_MART_FUNCGEN Ensembl Regulation 100</pre></body></html></div>
</div>
<div id="biomart-data-sets" class="section level2">
<h2 class="hasAnchor">
<a href="#biomart-data-sets" class="anchor"></a>biomaRt data sets</h2>
<p>And we can then check for the available data sets on a particular server.</p>
<div class="sourceCode" id="cb96"><html><body><pre class="r">mart &lt;- useEnsembl("ensembl")
head(listDatasets(mart))

                     dataset                                  description
1   acalliptera_gene_ensembl             Eastern happy genes (fAstCal1.2)
2 acarolinensis_gene_ensembl               Anole lizard genes (AnoCar2.0)
3  acchrysaetos_gene_ensembl              Golden eagle genes (bAquChr1.2)
4  acitrinellus_gene_ensembl               Midas cichlid genes (Midas_v5)
5  amelanoleuca_gene_ensembl                        Panda genes (ailMel1)
6    amexicanus_gene_ensembl Mexican tetra genes (Astyanax_mexicanus-2.0)
                 version
1             fAstCal1.2
2              AnoCar2.0
3             bAquChr1.2
4               Midas_v5
5                ailMel1
6 Astyanax_mexicanus-2.0
&gt; </pre></body></html></div>
</div>
<div id="biomart-queries" class="section level2">
<h2 class="hasAnchor">
<a href="#biomart-queries" class="anchor"></a>biomaRt queries</h2>
<p>After setting up a <code>mart</code> object pointing to the server and data set that we care about, we can make queries. We first set up the <code>mart</code> object.</p>
<div class="sourceCode" id="cb97"><html><body><pre class="r"><span class="no">mart</span> <span class="kw">&lt;-</span> <span class="fu">useEnsembl</span>(<span class="st">"ensembl"</span>,<span class="st">"hsapiens_gene_ensembl"</span>                   )</pre></body></html></div>
<p>Queries are of the form</p>
<p>getBM(<em>attributes</em>, <em>filters</em>, <em>values</em>, <em>mart</em>)</p>
<p>where</p>
<ul>
<li><p>attributes are the things we <strong>want</strong></p></li>
<li><p>filters are the <em>types of</em> IDs we <strong>have</strong></p></li>
<li><p>values are the IDs we <strong>have</strong></p></li>
<li><p>mart is the <code>mart</code> object we set up</p></li>
</ul>
</div>
<div id="biomart-attributes-and-filters" class="section level2">
<h2 class="hasAnchor">
<a href="#biomart-attributes-and-filters" class="anchor"></a>biomaRt attributes and filters</h2>
<p>Both attributes and filters have rather inscrutable names, but a listing can be accessed using</p>
<div class="sourceCode" id="cb98"><html><body><pre class="r">atrib &lt;- listAttributes(mart)
filts &lt;- listFilters(mart)
head(atrib)
                           name                  description         page
1               ensembl_gene_id               Gene stable ID feature_page
2       ensembl_gene_id_version       Gene stable ID version feature_page
3         ensembl_transcript_id         Transcript stable ID feature_page
4 ensembl_transcript_id_version Transcript stable ID version feature_page
5            ensembl_peptide_id            Protein stable ID feature_page
6    ensembl_peptide_id_version    Protein stable ID version feature_page
head(filts)
                name                            description
1    chromosome_name               Chromosome/scaffold name
2              start                                  Start
3                end                                    End
4             strand                                 Strand
5 chromosomal_region e.g. 1:100:10000:-1, 1:100000:200000:1
6          with_ccds                        With CCDS ID(s)</pre></body></html></div>
</div>
<div id="biomart-query" class="section level2">
<h2 class="hasAnchor">
<a href="#biomart-query" class="anchor"></a>biomaRt query</h2>
<p>A simple example query</p>
<div class="sourceCode" id="cb99"><html><body><pre class="r">afyids &lt;- c("1000_at","1001_at","1002_f_at","1007_s_at")
getBM(c("affy_hg_u95av2", "hgnc_symbol"), c("affy_hg_u95av2"), afyids, mart)

  affy_hg_u95av2 hgnc_symbol
1        1000_at       MAPK3
2      1007_s_at        DDR1
3        1001_at        TIE1
4      1002_f_at            
5      1002_f_at     CYP2C19</pre></body></html></div>
</div>
<div id="biomart-exercises" class="section level2">
<h2 class="hasAnchor">
<a href="#biomart-exercises" class="anchor"></a>biomaRt exercises</h2>
<ul>
<li><p>Get the Ensembl gene IDs and HUGO symbol for Entrez Gene IDs 672, 5468 and 7157</p></li>
<li><p>What do you get if you query for the ‘gene_exon’ for GAPDH?</p></li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by James MacDonald, Lori Shepherd.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
